cmd/globaldns.go

```diff
@@ -0,0 +1,865 @@
+
+	if providerType == "" {
+		return errors.New("--type is required")
+	}
+
+	provider := &managementClient.GlobalDNSProvider{
+		Name:       name,
+		RootDomain: rootDomain,
+	}
+	switch providerType {
+	case route53Provider:
+		accessKey := os.Getenv(envAwsAccessKeyID)
+		if accessKey == "" {
+			return errors.New("missing access key, please provide it via AWS_ACCESS_KEY_ID environment variable")
+		}
+		secretKey := os.Getenv(envAwsSecretAccessKey)
+		if secretKey == "" {
+			return errors.New("missing secret key, please provide it via AWS_SECRET_ACCESS_KEY environment variable")
+		}
+		provider.Route53ProviderConfig = &managementClient.Route53ProviderConfig{
+			AccessKey: accessKey,
+			SecretKey: secretKey,
+		}
+	case cloudflareProvider:
+		apiEmail := os.Getenv(envCfAPIEmail)
+		if apiEmail == "" {
+			return errors.New("missing API email, please provide it via CF_API_EMAIL environment variable")
+		}
+		apiKey := os.Getenv(envCfAPIKey)
+		if apiKey == "" {
+			return errors.New("missing API key, please provide it via CF_API_KEY environment variable")
+		}
+		provider.CloudflareProviderConfig = &managementClient.CloudflareProviderConfig{
+			APIEmail: apiEmail,
+			APIKey:   apiKey,
+		}
+	case alidnsProvider:
+		accessKey := os.Getenv(envAccessKeyID)
+		if accessKey == "" {
+			return errors.New("missing access key, please provide it via ACCESS_KEY_ID environment variable")
+		}
+		secretKey := os.Getenv(envAccessKeySecret)
+		if secretKey == "" {
+			return errors.New("missing secret key, please provide it via ACCESS_KEY_SECRET environment variable")
+		}
+		provider.AlidnsProviderConfig = &managementClient.AlidnsProviderConfig{
+			AccessKey: accessKey,
+			SecretKey: secretKey,
+		}
+	default:
+		return fmt.Errorf("unsupported provider type %q", providerType)
+	}
+
+	members, err := addMembersByNames(ctx, c, provider.Members, ctx.StringSlice(argMember), memberAccessTypeOwner)
+	if err != nil {
+		return err
+	}
+	provider.Members = members
+
+	if _, err = c.ManagementClient.GlobalDNSProvider.Create(provider); nil != err {
+		return err
+	}
+
+	fmt.Printf("Successfully created global DNS provider %q\n", name)
+	return nil
+}
+
+func globalDNSProviderUpdate(ctx *cli.Context) error {
+	if ctx.NArg() == 0 {
+		return cli.ShowSubcommandHelp(ctx)
+	}
+
+	c, err := GetClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	name := ctx.Args().First()
+
+	provider, err := searchForGlobalDNSProvider(c, name)
+	if err != nil {
+		return err
+	}
+
+	update := make(map[string]interface{})
+
+	if ctx.IsSet(argRootDomain) {
+		update["rootDomain"] = ctx.String(argRootDomain)
+	}
+
+	providerType := ""
+	if provider.Route53ProviderConfig != nil {
+		providerType = route53Provider
+	} else if provider.CloudflareProviderConfig != nil {
+		providerType = cloudflareProvider
+	} else if provider.AlidnsProviderConfig != nil {
+		providerType = alidnsProvider
+	}
+
+	switch providerType {
+	case route53Provider:
+		accessKey := os.Getenv(envAwsAccessKeyID)
```

Comment:
Should we have a way to get those parameters from cli flags instead of getting them from environment variable?

Reply:
like we can support --args AWS_ACCESS_KEY_ID=xxxxxxx --args AWS_SECRET_ACCESS_KEY=xxxx

Reply:
Updated